### PR TITLE
1799: Update indicator tests

### PR DESCRIPTION
**Pull request by kael89**
### Issue #:
Adds integration tests in `indicators` + updates unit tests for https://github.com/beyondessential/tupaia-backlog/issues/1799

#### Integration test conventions

This PR introduces some conventions for integration tests. I'm happy to discuss their implementation or whether they are beneficial at all!

Integration tests are:
* Placed under an `__integration` folder inside `__tests__`. We could also name this folder `__integration__`
* Config driven - it should be fairly easy to add a new test case
* Mocking should be minimal: network requests to 3rd party APIs, and potentially datetime functionality
* The input and output should be very specific
* They generally test realistic "happy paths" and main error throwing - unit tests can be used to test edge cases (there can be exceptions to this in practice)

A lot of the above are already implied in existing integration tests in our codebase. I used a specific approach here:
* A `.json` file serves as the test config. `.js` files may abstract the config generation, but in this case it's probably easier to use plain config. A drawback is that you can't have comments in `.json`, so we could also try `.jsonc`
* A generic `runItntegrationTest` helper that encourages adhering to the above by standardising test aspects (setup, input, expected output)

My intention would be for the same or similar pattern to be used for API testing (like the one we do in `meditrak-server` currently, using supertest). The used endpoints and expected responses can easily be specified in config. One difference here is that we may choose to place those under an `__api` folder.

### CODE REVIEW

File: packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+}));
+
+describe('ArithmeticBuilder', () => {
+  describe('validateConfig()', testConfigValidation);
+
+  describe('getAggregations()', testGetAggregations);
+
+  describe('getElementCodes()', testGetElementCodes);
+
+  describe('buildAnalytics()', testBuildAnalytics);
```
Comment (kael89) [inline packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts:25]: Apart from `validateConfig`, all other unit tests here explore aspects that are already semi-tested in integration tests. I thought that it's still probably useful (and "better practice"?) to keep the unit tests, since they make it easy to define and test more elaborate scenarios/edge cases.

This decision was made easier by the fact that most of them were already written. I don't know if I would go this way if I had started by writing integration tests in this package. When to use each category seems like is an interesting problem to me, and probably has no specific answers. I'd be happy to hear your thoughts!

Comment (kael89) [reply packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts:25]: Also, we don't want to "over-test" a package since tests have maintenance overhead. I tried to be conscious of this but please let me know if you think this is the case here!

For future reference: integration tests probably have minimal maintenance cost and I wish I had them written before starting this refactor! Unit tests required some time to be updated but not too much.

File: packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+          "startDate": "2019-01-01",
+          "endDate": "2019-01-31",
+          "organisationUnitCodes": ["GR"]
+        }
+      },
+      "expected": [
+        {
+          "dataElement": "MonthlyBirths",
+          "organisationUnit": "GR",
+          "period": "20190101",
```
Comment (kael89) [inline packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:240]: This doesn't seem correct, right? `MonthlyBirths` uses a `FINAL_EACH_MONTH` aggregation so one would expect this to be `201901`!

I believe this is a "bug" in this aggregation - as a matter of fact, most current aggregations almost never concern themselves with normalising the return period based on what they do. We'd need to change that to make them compatible with the indicator architecture - probably in https://github.com/beyondessential/tupaia-backlog/issues/2150

We were lucky in `PSSS`: we use week types there which force a period update (periods include a `W` in them so the conversion is inevitable for inclusion/comparison purposes etc)

File: packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+        {
+          "dataElement": "AnnualBirths",
+          "organisationUnit": "GR",
+          "period": "2020",
+          "value": 21
+        }
+      ]
+    },
+    {
+      "description": "Applies correct aggregation (object format)",
```
Comment (kael89) [inline packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:328]: This test case will fail - and that's a good thing!! Again, I believe we got lucky in PSSS - there are a few bugs related to `fetchOptions` expansion during fetching, and during aggregation. Good news - this is also related to your original comment in the previous PR: https://github.com/beyondessential/tupaia/pull/1873#issuecomment-761892449.

I'll submit a new PR to fix this, and may add a couple of more complicated test cases here (nested indicators, including `OFFSET_PERIOD` and `SUM` aggregations, make sure these work correctly together)

**Comment by edmofro** [comment]
Is it expected that this should fail on codeship @kael89?

File: packages/indicators/src/types.ts [hunk 4]
```diff
@@ -21,50 +21,58 @@ export interface AnalyticCluster {
   dataValues: Record<Analytic['dataElement'], Analytic['value']>;
 }
 
-type TypeFields = Record<string, string | number | Record<string, unknown>>;
+type DbValue = string | number | boolean | null | DbValue[] | { [key: string]: DbValue };
```
Comment (edmofro) [inline packages/indicators/src/types.ts:24]: Some intense type nesting here, I'm impressed TypeScript is happy to evaluate that!

File: packages/indicators/src/__tests__/__integration/integration.test.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    indicatorCodes: string[];
+    fetchOptions: FetchOptions;
+  };
+  expected: Analytic[] | string;
+  throws?: boolean;
+}
+
+const dataBroker = new DataBroker();
+// TODO Use the base `@tupaia/database` models instead of `dataBroker.models`
+// after https://github.com/beyondessential/tupaia-backlog/issues/663 is implemented
```
Comment (edmofro) [inline packages/indicators/src/__tests__/__integration/integration.test.ts:24]: If this is because the `findOrDefault` behaviour is needed, I believe you pulled that into the central model recently?

File: packages/indicators/src/__tests__/Builder/ArithmeticBuilder/testBuildAnalytics.ts [hunk 6]
```diff
@@ -71,35 +72,31 @@ export const testCalculations = () => {
...
         {
           formula: 'A_ToPg_20192020 + B_ToPg_20192020',
           aggregation: 'FINAL_EACH_YEAR',
         },
         [
-          { organisationUnit: 'TO', period: '2019', value: 3 },
-          { organisationUnit: 'TO', period: '2020', value: 30 },
-          { organisationUnit: 'PG', period: '2019', value: 3.3000000000000003 },
-          { organisationUnit: 'PG', period: '2020', value: 33 },
+          ['TO', '2019', 3],
```
Comment (edmofro) [inline packages/indicators/src/__tests__/Builder/ArithmeticBuilder/testBuildAnalytics.ts:96]: This is a nice example of when the values are self describing enough, that the field names really are just clutter - I like the array format :-)

**Comment by kael89** [comment]
> Is it expected that this should fail on codeship @kael89?

Yes - see my [comment](https://github.com/beyondessential/tupaia/pull/1919#discussion_r560586349)

File: packages/database/src/testUtilities/runIntegrationTests.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * @property {Object<string, *>} config
+ * @property {Function} [beforeAll]
+ * @property {Function} [afterAll]
+ * @property {Function} it
+ */
+
+/**
+ * @param {IntegrationTestInput} input
+ */
+export const runIntegrationTests = ({
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/testUtilities/runIntegrationTests.js:29]: Hmmmm...good question! I don't have a strong opinion. It feels wrong here, but I totally see your point about introducing the `database` dependency into `utils`. Your call :-)

File: packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import { testBuildAnalytics } from './testBuildAnalytics';
+import { testGetAggregations } from './testGetAggregations';
+import { testGetElementCodes } from './testGetElementCodes';
+import { testConfigValidation } from './testValidateConfig';
+
+jest.mock('../../../Builder/createBuilder', () => ({
+  createBuilder: jest.fn().mockImplementation((indicator: Indicator) => ({
+    getElementCodes: () => indicator.config.elementCodes,
+  })),
+}));
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts:16]: Yeah ... that's a shame. Could you put a comment in `testGetElementCodes` to at least point people in the right direction if they're looking?

File: packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts (hunk 1 above)
Comment (edmofro) [reply packages/indicators/src/__tests__/Builder/ArithmeticBuilder/ArithmeticBuilder.test.ts:25]: I'm keen to keep both, seeing as we have them. In the end, if we find ourselves spending inordinate amounts of time updating unit tests, we can kill them at that point!

For the general case - yeah I agree that there's no definite answer to when one should definitely be used over the other. I favour integration tests for "bang for buck", but unit tests are ideal in situations when there's a specific piece of logic that lives in a single place and is complex/risky/central to the system's successful functioning.

File: packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+          "endDate": "2019-01-31",
+          "organisationUnitCodes": ["GR"]
+        }
+      },
+      "expected": [
+        {
+          "dataElement": "FemaleGtMale",
+          "organisationUnit": "GR",
+          "period": "20190101",
+          "value": 0
```
Comment (edmofro) [inline packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:260]: So `true` is `0`, and `false` `1`?

File: packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json (hunk 3 above)
Comment (edmofro) [reply packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:328]: Oh ok cool! Nice test case :-)

**Review by edmofro — COMMENTED** [review]
This is an awesome setup for integration tests!! 🎉 

I'm not totally convinced one big JSON file is the most clear way to define them - personally I'd prefer being able to structure things a little more explicitly. E.g. naming the input fixture blocks individually, then pulling them all together, and having that separate to the test cases. However, overall I think this is a big win to allow us to have consistent and declarative creation of integration tests. Have you played around with the same sort of thing defined in a `.js` file?

Also, I'm not going to pretend I checked all the logic of the test cases against the data fixtures - in most cases (particularly within the integration tests) I just trusted the test case descriptions.

File: packages/indicators/src/types.ts (hunk 4 above)
Comment (kael89) [reply packages/indicators/src/types.ts:24]: Yeah TypeScript is very sophisticated!

File: packages/indicators/src/__tests__/__integration/integration.test.ts (hunk 5 above)
Comment (kael89) [reply packages/indicators/src/__tests__/__integration/integration.test.ts:24]: Correct! And I'm about to remove `findOrDefault` completely after implementing https://github.com/beyondessential/tupaia-backlog/issues/663, so I 'll update this after 663 ends in `dev`, which should be soon enough

File: packages/indicators/src/__tests__/Builder/ArithmeticBuilder/testBuildAnalytics.ts (hunk 6 above)
Comment (kael89) [reply packages/indicators/src/__tests__/Builder/ArithmeticBuilder/testBuildAnalytics.ts:96]: Same here! The object format seems handy for passing analytics around, but array seems simpler for defining them!

File: packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json (hunk 9 above)
Comment (kael89) [reply packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:260]: Umm no actually this is an error in the test case description!

Perhaps following the feminist tradition in Tupaia I subconsciously wanted the `FemaleGtMale` statement to be `true` rather than false 😉

Comment (edmofro) [reply packages/indicators/src/__tests__/__integration/arithmeticAnalytics.fixtures.json:260]: Ah right! Haha nice, Elena would be proud

File: packages/database/src/testUtilities/upsertDummyRecord.js [hunk 10]
```diff
@@ -74,3 +74,11 @@ export const findOrCreateDummyRecord = async (model, findCriteria, data) => {
   const generatedData = await generateDummyRecord(model, { ...findCriteria, ...data });
   return model.findOrCreate(findCriteria, generatedData);
 };
+
+export const findOrCreateRecords = async (models, recordsByType) => {
```
[the remark this answers is not in the document]
Comment (kael89) [reply packages/database/src/testUtilities/upsertDummyRecord.js:78]: I've thought the same in the past, but when I tried to convert one to the other a few specific tests failed. One example (could be the only one!) is [GET /changes/count](https://github.com/beyondessential/tupaia/blob/dev/packages/meditrak-server/src/tests/routeTests/changes/get.js), where we want to make sure that a record has actually been upserted (and not "found"). Of course, we could ensure this in other ways (delete before insert/use a unique code etc).

In general, we have a lot of really handy test utils that have evolved a bit independently. Hopefully one day we'll bring all of them together in a standardised test DB setup system, monorepo-wide!

File: packages/database/src/testUtilities/upsertDummyRecord.js [hunk 11]
```diff
@@ -82,3 +82,17 @@ export const findOrCreateRecords = async (models, recordsByType) => {
     }
   }
 };
+
+/**
+ * Generates test data, and stores it in the database. Uses test ids so that all can be cleanly
+ * wiped afterwards. Any missing fields on the records passed in are generated randomly or using
+ * sensible defaults, using the logic in upsertDummyRecord
+ */
+export const populateTestData = async (models, recordsByType) => {
```
Comment (edmofro) [inline packages/database/src/testUtilities/upsertDummyRecord.js:91]: Thanks for moving this alongside the others 🙌

**Comment by kael89** [comment]
> This is an awesome setup for integration tests!! 🎉
> 
> I'm not totally convinced one big JSON file is the most clear way to define them - personally I'd prefer being able to structure things a little more explicitly. E.g. naming the input fixture blocks individually, then pulling them all together, and having that separate to the test cases. However, overall I think this is a big win to allow us to have consistent and declarative creation of integration tests. Have you played around with the same sort of thing defined in a `.js` file?

Thanks @edmofro, good points about defining integration tests setup in multiple `.js` files. I tried it briefly and keen to give it a second shot, after all required code fixes are done.

**Review by edmofro — APPROVED** [review]
Would still be keen to check out the readability of js files instead, but that aside this is all approved 🎉

**Comment by edmofro** [comment]
> > This is an awesome setup for integration tests!! 🎉
> > I'm not totally convinced one big JSON file is the most clear way to define them - personally I'd prefer being able to structure things a little more explicitly. E.g. naming the input fixture blocks individually, then pulling them all together, and having that separate to the test cases. However, overall I think this is a big win to allow us to have consistent and declarative creation of integration tests. Have you played around with the same sort of thing defined in a `.js` file?
> 
> Thanks @edmofro, good points about defining integration tests setup in multiple `.js` files! I tried it briefly and keen to give it a second shot, after all required code fixes are done.

Cools, and I actually wasn't meaning multiple `.js` files - even all in one file, but the js syntax could allow better grouping and visual discrimination of sections. Let me know how it goes :-)
